app/src/main/java/ca/bc/gov/bchealth/ui/healthrecord/labtest/LabTestDetailFragment.kt

```diff
@@ -76,21 +89,85 @@ class LabTestDetailFragment : Fragment(R.layout.fragment_lab_test_detail) {
 
                     if (state.onError) {
                         showError()
+                        viewModel.resetUiState()
                     }
+
+                    handlePdfDownload(state)
                 }
             }
         }
     }
 
+    private fun handlePdfDownload(state: LabTestDetailUiState) {
+        if (state.showDownloadOption) {
+            binding.toolbar.ivRightOption.isVisible = true
+        }
+
+        if (state.pdfData?.isNotEmpty() == true) {
+            pdfDecoderViewModel.base64ToPDFFile(state.pdfData)
+            viewModel.resetUiState()
+        }
+    }
+
     private fun showError() {
         AlertDialogHelper.showAlertDialog(
             context = requireContext(),
             title = getString(R.string.error),
             msg = getString(R.string.error_message),
-            positiveBtnMsg = getString(R.string.dialog_button_ok),
-            positiveBtnCallback = {
-                findNavController().popBackStack()
+            positiveBtnMsg = getString(R.string.dialog_button_ok)
+        )
+    }
+
+    private fun observePdfData() {
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                pdfDecoderViewModel.uiState.collect { uiState ->
+                    uiState.pdf?.let {
+                        val (base64Pdf, file) = it
+                        if (file != null) {
+                            try {
+                                fileInMemory = file
+                                showPDF(file)
+                            } catch (e: Exception) {
+                                fallBackToPdfRenderer(base64Pdf)
+                            }
+                        } else {
+                            fallBackToPdfRenderer(base64Pdf)
+                        }
+                        pdfDecoderViewModel.resetUiState()
+                    }
+                }
             }
+        }
+    }
+
+    private fun showPDF(file: File) {
```

Comment:
Can you add this method to common file. So it will remove code duplication.

Reply:
Resolved this by using helper class to show PDF.